File: src/audio_stegano_app/lib/core/stego/cover_record_budget.dart
/// Minimum cover-recording budget so LSB capacity ≥ [requiredBits]
/// (one bit per mono sample, same as [EmbedMessage] capacity).
///
/// Prefer [samplesSatisfied] / [progressFromSamples] — wall-clock alone is
/// unreliable on web (undersampled streams vs nominal [coverSampleRate]).
class CoverRecordBudget {
  CoverRecordBudget._();

  static const int coverSampleRate = 44100;

  /// Extra mono samples beyond the bit budget (≈100 ms at 44.1 kHz).
  static const int safetySampleMargin = 4410;

  static int requiredSamples(int requiredBits) {
    if (requiredBits <= 0) return 0;
    return requiredBits + safetySampleMargin;
  }

  /// Wall-clock estimate with safety factor (UI hint only — not the stop gate).
  static Duration minDuration({
    required int requiredBits,
    int sampleRate = coverSampleRate,
    double safetyFactor = 1.35,
  }) {
    if (requiredBits <= 0 || sampleRate <= 0) {
      return Duration.zero;
    }
    final samples = requiredSamples(requiredBits);
    final micros =
        ((samples * safetyFactor * 1000000) / sampleRate).ceil();
    return Duration(microseconds: micros);
  }

  static double progressFromSamples(int bufferedSamples, int requiredBits) {
    final need = requiredSamples(requiredBits);
    if (need <= 0) return 1;
    final p = bufferedSamples / need;
    if (p.isNaN || p.isInfinite) return 0;
    return p.clamp(0.0, 1.0);
  }

  static bool samplesSatisfied(int bufferedSamples, int requiredBits) =>
      bufferedSamples >= requiredSamples(requiredBits);

  /// Remaining wall-clock estimate from missing samples at [sampleRate].
  static Duration remainingFromSamples({
    required int bufferedSamples,
    required int requiredBits,
    int sampleRate = coverSampleRate,
  }) {
    final need = requiredSamples(requiredBits) - bufferedSamples;
    if (need <= 0 || sampleRate <= 0) return Duration.zero;
    final micros = ((need * 1000000) / sampleRate).ceil();
    return Duration(microseconds: micros);
  }

  @Deprecated('Use progressFromSamples')
  static double progress(Duration elapsed, Duration min) {
    if (min <= Duration.zero) return 1;
    final p = elapsed.inMicroseconds / min.inMicroseconds;
    if (p.isNaN || p.isInfinite) return 0;
    return p.clamp(0.0, 1.0);
  }

  @Deprecated('Use samplesSatisfied')
  static bool isSatisfied(Duration elapsed, Duration min) =>
      min <= Duration.zero || elapsed >= min;
}

File: src/audio_stegano_desktop/src/AudioStegano.Core/Stego/CoverRecordBudget.cs
namespace AudioStegano.Core.Stego;

/// <summary>
/// Minimum cover-recording budget so LSB capacity ≥ requiredBits (1 bit / mono sample).
/// Gate on buffered sample count — wall-clock alone can undershoot on slow capture.
/// </summary>
public static class CoverRecordBudget
{
    public const int CoverSampleRate = 44100;

    /// <summary>Extra mono samples beyond the bit budget (≈100 ms at 44.1 kHz).</summary>
    public const int SafetySampleMargin = 4410;

    public static int RequiredSamples(int requiredBits)
    {
        if (requiredBits <= 0) return 0;
        return requiredBits + SafetySampleMargin;
    }

    /// <summary>Wall-clock estimate with safety factor (UI hint only).</summary>
    public static TimeSpan MinDuration(
        int requiredBits,
        int sampleRate = CoverSampleRate,
        double safetyFactor = 1.35)
    {
        if (requiredBits <= 0 || sampleRate <= 0)
            return TimeSpan.Zero;
        var samples = RequiredSamples(requiredBits);
        return TimeSpan.FromSeconds(samples * safetyFactor / sampleRate);
    }

    public static double ProgressFromSamples(int bufferedSamples, int requiredBits)
    {
        var need = RequiredSamples(requiredBits);
        if (need <= 0) return 1;
        var p = bufferedSamples / (double)need;
        if (double.IsNaN(p) || double.IsInfinity(p)) return 0;
        return Math.Clamp(p, 0, 1);
    }

    public static bool SamplesSatisfied(int bufferedSamples, int requiredBits) =>
        bufferedSamples >= RequiredSamples(requiredBits);

    public static TimeSpan RemainingFromSamples(
        int bufferedSamples,
        int requiredBits,
        int sampleRate = CoverSampleRate)
    {
        var need = RequiredSamples(requiredBits) - bufferedSamples;
        if (need <= 0 || sampleRate <= 0) return TimeSpan.Zero;
        return TimeSpan.FromSeconds(need / (double)sampleRate);
    }
}
